### PR TITLE
Fix multi-interval series issue

### PR DESCRIPTION
**Pull request by shanson7**
Because we were not combining multiple intervals for the same series into a single idx.Node we were hitting issues with the code deduping the series on the remote end. This change should fix both cases:

multiple defs on a single node should be collapsed into a single idx.Node element
multiple defs across nodes should not overwrite one another and should be merged later (by mergeSeries)

The symptom was data sort of "flashing" on refresh, switching between the old and new intervals data (since only one was winning).

### CODE REVIEW

**Comment by Dieterbe** [comment]
> Because we were not combining multiple intervals for the same series into a single idx.Node 

I'm missing a lot of info here. we = your deployment? or we = the metrictank codebase? 
FindByTag returns a idx.Node for each MKey,
Find returns an idx.Node for each path (nameWithTags)
so did you modify the code? were you creating new paths when you change the interval?
can you describe the problem in more detail?

**Comment by shanson7** [comment]
> can you describe the problem in more detail?

Sure. [FindByTag](https://github.com/grafana/metrictank/blob/5a8688e2bfcad6996831b9411ea54c5dad3d15b8/idx/memory/memory.go#L844) will always return exactly one def per `idx.Node`. If there are multiple `idx.Node`s with the same path, one of them is picked arbitrarily by [this logic](https://github.com/grafana/metrictank/blob/5a8688e2bfcad6996831b9411ea54c5dad3d15b8/api/graphite.go#L969), which will cause a series with multiple intervals to be incomplete.

As an example, if you had `foo.bar;baz=v` and it was published at a 30s interval and then changed to a 1min interval, *both* would be returned as separate `idx.Node` from `FindByTag` and `clusterFindyByTag` would only keep 1 of them.  In practice, this means that graphing this series over a time range that contains both series would cause "flapping" where one would show, then the other.

**Comment by Dieterbe** [comment]
Now your explanation makes a lot more sense.
And for comparison, MemoryIdx.Find (or more specifically MemoryIdx.find) returns idx.Node which has multiple Defs (e.g. for multiple intervals) under the same path (that's how we store in memory  because Tree.Items holds the memory.Node and Node has mutiple Defs). I just tested and confirms this all still works.

I agree that we should fix this, and generally we should make FindByTag more symmetrical to Find.

File: idx/memory/memory.go [hunk 1]
```diff
@@ -841,12 +847,27 @@ func (m *MemoryIdx) FindByTag(orgId uint32, expressions []string, from int64) ([
 			continue
 		}
 
-		res = append(res, idx.Node{
-			Path:        def.NameWithTags(),
-			Leaf:        true,
-			HasChildren: false,
-			Defs:        []idx.Archive{*def},
-		})
+		if _, ok := seen[def.NameWithTags()]; !ok {
```
[the remark this answers is not in the document]
Comment (shanson7) [reply idx/memory/memory.go]: Good point. I had actually implemented it like this originally, but found the `defByTagSet` and went that path. I didn't think about this though. Will fix

[the remark this answers is not in the document]
Comment (Dieterbe) [reply idx/memory/memory.go]: i agree. irrespective of the lastUpdate check (which indeed can be removed) that seems simpler, cleaner and also more symmetrical to what MemoryIdx.Find() does

File: idx/memory/memory.go [hunk 2]
```diff
@@ -222,7 +222,10 @@ func (m *MemoryIdx) Update(point schema.MetricPoint, partition int32) (idx.Archi
 		if LogLevel < 2 {
 			log.Debug("metricDef with id %v already in index", point.MKey)
 		}
-		existing.LastUpdate = int64(point.Time)
+
+		if existing.LastUpdate < int64(point.Time) {
+			existing.LastUpdate = int64(point.Time)
+		}
```
[the remark this answers is not in the document]
Comment (Dieterbe) [reply idx/memory/memory.go]: very interesting. note that the tank (AggMetric) wouldn't add this point (unless it is accepted by the reorder buffer), which makes this case even more interesting.
Also this only seems a problem if this "buggy, old" point is not followed by a correct, recent point.
you're the first one I know with a stream this broken, but sure, we may as well protect against it :)

[the remark this answers is not in the document]
Comment (Dieterbe) [reply idx/memory/memory.go]: BTW, comparing Find with FindByTag and checking for how they differ is an interesting exercise. I found another bug which we can fix later: #899
probably no one has noticed cause no one uses the feature (except worldping which is currently untagged)

**Comment by shanson7** [comment]
I'm not sure why a test failed here. It doesn't look like it's due to my change, but I cannot retry it.

**Comment by Dieterbe** [comment]
@shanson7 this PR looks :+1: i just added a few minor cleanups and will merge shortly

**Pull request merged**